wallet/import.go

```diff
@@ -1,13 +1,191 @@
+
+	// Private keys are not allowed.
+	if pubKey.IsPrivate() {
+		return errors.New("private keys cannot be imported")
+	}
+
+	// Public key must have a version corresponding to the current chain.
+	if !pubKey.IsForNet(w.chainParams) {
+		return fmt.Errorf("expected extended public key for current "+
+			"network %v", w.chainParams.Name)
+	}
+
+	// Verify the extended public key's depth and child index based on
+	// whether it's a master key or not.
+	if isMasterKey {
+		if pubKey.Depth() != masterPubKeyDepth {
+			return errors.New("invalid master key, must be of the " +
+				"form m/purpose'/coin_type'/account'")
+		}
+		if pubKey.ChildIndex() < hdkeychain.HardenedKeyStart {
+			return errors.New("invalid master key, must be hardened")
+		}
+	} else {
+		if pubKey.Depth() != pubKeyDepth {
+			return errors.New("invalid master key, must be of the " +
+				"form m/purpose'/coin_type'/account'/change/" +
+				"address_index")
+		}
+		if pubKey.ChildIndex() >= hdkeychain.HardenedKeyStart {
+			return errors.New("invalid pulic key, must not be " +
+				"hardened")
+		}
+	}
+
+	return nil
+}
+
+// ImportAccount imports an account backed by a master extended public key. The
+// master key fingerprint denotes the fingerprint of the root key corresponding
+// to the master public key (also known as the key with derivation path m/).
+// This may be required by some hardware wallets for proper identification and
+// signing. The address type can usually be inferred from the key's version, but
+// in the case of legacy versions (xpub, tpub), an address type must be
+// specified as we intend to not support importing BIP-44 keys into the wallet
+// using the legacy pay-to-pubkey (P2PK) scheme.
+func (w *Wallet) ImportAccount(name string, masterPubKey *hdkeychain.ExtendedKey,
+	masterKeyFingerprint uint32, addrType *waddrmgr.AddressType) error {
+
+	// Ensure we have a valid master public key.
+	if err := w.validateExtendedPubKey(masterPubKey, true); err != nil {
+		return err
+	}
+
+	// Determine what key scope the master public key should belong to and
+	// init the account as watch-only.
+	keyScope, err := keyScopeFromPubKey(masterPubKey, addrType)
+	if err != nil {
+		return err
+	}
+	scopedMgr, err := w.Manager.FetchScopedKeyManager(keyScope)
+	if err != nil {
+		return err
+	}
+
+	return walletdb.Update(w.db, func(tx walletdb.ReadWriteTx) error {
+		ns := tx.ReadWriteBucket(waddrmgrNamespaceKey)
+		_, err = scopedMgr.NewAccountWatchingOnly(
+			ns, name, masterPubKey, masterKeyFingerprint,
+		)
+		return err
+	})
+}
+
+// ImportPubKey imports a single derived public key into the address manager.
+// The address type can usually be inferred from the key's version, but in the
+// case of legacy versions (xpub, tpub), an address type must be specified as we
+// intend to not support importing BIP-44 keys into the wallet using the legacy
+// pay-to-pubkey (P2PK) scheme.
+func (w *Wallet) ImportPubKey(pubKey *hdkeychain.ExtendedKey,
```

Comment:
If this is a plain public key, do we actually need the extra information in the `ExtendedKey` struct?

Reply:
Not really, I went with it here to keep things consistent. Should we change it?

Reply:
Hmm, no strong prefrecne, but IMO the API woul dbe more self documenting if it was just a regular public key, given we don't use any of the other information that's carried along with an extended pubkey.

Reply:
Converted to use standard public keys.